base/rational.jl

```diff
@@ -140,10 +140,21 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
 (::Type{T})(x::Rational) where {T<:Integer} = (isinteger(x) ? convert(T, x.num)::T :
     throw(InexactError(nameof(T), T, x)))
 
-AbstractFloat(x::Rational) = (float(x.num)/float(x.den))::AbstractFloat
-function (::Type{T})(x::Rational{S}) where T<:AbstractFloat where S
-    P = promote_type(T,S)
-    convert(T, convert(P,x.num)/convert(P,x.den))::T
+function numerator_denominator_promoted(x)
+    y = unsafe_rational(numerator(x), denominator(x))
+    (numerator(y), denominator(y))
+end
+
+function rational_to_floating_point(::Type{F}, x, rm = RoundNearest, prec = precision(F)) where {F}
```

Comment:
do we actually care about giving the user the ability to specify rounding modes here? Most other functions in `Base` don't, and giving up that seems like it would simplify the code a lot.

Reply:
We need the rounding modes for `BigFloat`. Also, having rounding modes other than the default one enables more thorough testing that applies to all rounding modes.

Reply:
Most of the logic for the rounding modes is now in rounding.jl anyway.

Reply:
`to_floating_point_impl(::Type{T}, ::Type{S}, num, den, rm, prec) where {T<:Base.IEEEFloat,S}` could be simplified by requiring `rm == RoundNearest`, but then the tests would suffer.

Reply:
> giving the user the ability

to be clear, that particular method is not meant to be public

Reply:
Can't we do the `BigFloat` ones by just doing arbitrary precision `BigFloat` division? I'm not really sure if it's worth adding a bunch of code to make `BigFloat(::Rational)` fast.

Reply:
I don't think that would be an improvement, by any metric. The relevant code was already added in #50691 (into rounding.jl), so now it doesn't cost anything to support the known rounding modes for `BigFloat`, although the `IEEEFloat` method could be simplified a bit if necessary, at the cost of some testing, as mentioned above.

Reply:
> by any metric

Maybe the way you and @Joel-Dahne propose would be faster, though? But still, having everything under a single code path seems better as any bugs should manifest (and be fixed) sooner, so it's more maintainable.